clang/lib/Sema/TreeTransform.h

```diff
@@ -12864,6 +12864,19 @@ TreeTransform<Derived>::TransformCXXNewExpr(CXXNewExpr *E) {
     ArraySize = NewArraySize.get();
   }
 
+  // Per C++0x [expr.new]p5, the type being constructed may be a
+  // typedef of an array type.
+  QualType AllocType = AllocTypeInfo->getType();
+  if (ArraySize && E->isTypeDependent()) {
```

Comment:
does Value dependence matter?

Reply:
Yes, it does. Both the original testcase from @slackito  and simpler version of @dwblaikie are working now. 
normal arrays with new initializer also were passing the check and their type was correct earlier so need for modification for them. ```typedef TYPE TArray[8]; TArray x;``` here type of array was Tarray which should be Type* so the check only for type dependent case.

Reply:
Note I asked about VALUE dependence, not TYPE dependence.  I see why type dependence would matter, but expressions can be both type and value dependent.

Reply:
No, it doesn't have value dependence.